### ISSUE TITLE
Sik-Ho Tsang | Review -- GhostNet: More Features from Cheap Operations.

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#77]
Sik-Ho Tsang. [Review — GhostNet: More Features from Cheap Operations](https://sh-tsang.medium.com/review-ghostnet-more-features-from-cheap-operations-1784f3bbc2b).

**Comment by NorbertZheng** [comment]
## Overview
![image](https://user-images.githubusercontent.com/37235023/229402817-7aeeef44-6c24-41c7-a5f7-035e8d2ec8af.png)
Visualization of some feature maps generated by the first residual group in [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8?source=post_page---------------------------)-50, where three similar feature map pair examples are annotated with boxes of the same color. One feature map in the pair can be approximately obtained by transforming the other one through cheap operations (denoted by spanners).

[GhostNet: More Features from Cheap Operations](https://openaccess.thecvf.com/content_CVPR_2020/papers/Han_GhostNet_More_Features_From_Cheap_Operations_CVPR_2020_paper.pdf), GhostNet, by Huawei Technologies, Peking University, and University of Sydney, 2020 CVPR, Over 1100 Citations.

Image Classification
1989 … 2022 [[ConvNeXt](https://sh-tsang.medium.com/review-convnext-a-convnet-for-the-2020s-53b9ada30ab9)] [[PVTv2](https://sh-tsang.medium.com/review-pvtv2-improved-baselines-with-pyramid-vision-transformer-5cfd354a53bd)] [[ViT-G](https://sh-tsang.medium.com/review-scaling-vision-transformers-fc11d867ada6)] [[AS-MLP](https://sh-tsang.medium.com/review-as-mlp-an-axial-shifted-mlp-architecture-for-vision-a59c7fcbb3e6)] [[ResTv2](https://sh-tsang.medium.com/brief-review-rest-v2-simpler-faster-and-stronger-8de1ec5c0)] [[CSWin Transformer](https://sh-tsang.medium.com/review-cswin-transformer-a-general-vision-transformer-backbone-with-cross-shaped-windows-4cb4eb6e9ff5)] [[Pale Transformer](https://sh-tsang.medium.com/review-pale-transformer-a-general-vision-transformer-backbone-with-pale-shaped-attention-33bae6ae0571)] [[Sparse MLP](https://sh-tsang.medium.com/review-sparse-mlp-for-image-recognition-is-self-attention-really-necessary-60bd40d49e0a)] [[MViTv2](https://medium.com/p/a13b92ad3b37)] [[S²-MLP](https://sh-tsang.medium.com/review-s%C2%B2-mlp-spatial-shift-mlp-architecture-for-vision-224c1c7f2137)] [[CycleMLP](https://sh-tsang.medium.com/review-cyclemlp-a-mlp-like-architecture-for-dense-prediction-efae51ebb651)] [[MobileOne](https://sh-tsang.medium.com/review-mobileone-an-improved-one-millisecond-mobile-backbone-56b5cbe26a93)] [[GC ViT](https://sh-tsang.medium.com/review-gc-vit-global-context-vision-transformer-b220a97b5c64)] [[VAN](https://sh-tsang.medium.com/review-van-visual-attention-network-8e0218b8732e)] [[ACMix](https://sh-tsang.medium.com/review-acmix-on-the-integration-of-self-attention-and-convolution-c594bba6d494)] [[CVNets](https://sh-tsang.medium.com/review-cvnets-high-performance-library-for-computer-vision-954890bb7033)] [[MobileViT](https://sh-tsang.medium.com/review-mobilevit-light-weight-general-purpose-and-mobile-friendly-vision-transformer-3ef154376e93)] [[RepMLP](https://sh-tsang.medium.com/review-repmlp-re-parameterizing-convolutions-into-fully-connected-layers-for-image-recognition-1433cf9f82a6)] [[RepLKNet](https://sh-tsang.medium.com/review-scaling-up-your-kernels-to-31x31-revisiting-large-kernel-design-in-cnns-f10de3af873b)] [[ParNet](https://sh-tsang.medium.com/brief-review-parnet-non-deep-networks-e2e1d0d243aa)] 2023 [[Vision Permutator (ViP)](https://sh-tsang.medium.com/review-vision-permutator-a-permutable-mlp-like-architecture-for-visual-recognition-283ed62cf802)]
[==== My Other Paper Readings Are Also Over Here ====](https://sh-tsang.medium.com/overview-my-reviewed-paper-lists-tutorials-946ce59fbf9e)

 - A novel Ghost module is proposed to generate more feature maps from cheap operations, i.e.  enrich the latent space.
 - By stacking Ghost modules, GhostNet is formed.

**Comment by NorbertZheng** [comment]
## GhostNet
### Ghost Module
![image](https://user-images.githubusercontent.com/37235023/229403291-7403292e-c862-45b2-b68a-bc91057b028d.png)
An illustration of the convolutional layer and the proposed Ghost module for outputting the same number of feature maps. $\Phi$ represents the cheap operation.
 - **(a) Standard convolution.**
 - **(b) Ghost Module.** And the below source code is the ghost module implementation.

```python
class GhostModule(nn.Module):
    def __init__(self, inp, oup, kernel_size=1, ratio=2, dw_size=3, stride=1, relu=True):
        super(GhostModule, self).__init__()
        self.oup = oup
        init_channels = math.ceil(oup / ratio)
        new_channels = init_channels*(ratio-1)

        self.primary_conv = nn.Sequential(
            nn.Conv2d(inp, init_channels, kernel_size, stride, kernel_size//2, bias=False),
            nn.BatchNorm2d(init_channels),
            nn.ReLU(inplace=True) if relu else nn.Sequential(),
        )

        self.cheap_operation = nn.Sequential(
            nn.Conv2d(init_channels, new_channels, dw_size, 1, dw_size//2, groups=init_channels, bias=False),
            nn.BatchNorm2d(new_channels),
            nn.ReLU(inplace=True) if relu else nn.Sequential(),
        )

    def forward(self, x):
        x1 = self.primary_conv(x)
        x2 = self.cheap_operation(x1)
        out = torch.cat([x1,x2], dim=1)
        return out[:,:self.oup,:,:]
```

 - As seen from source code, standard convolution is performed for part of the feature maps, and $x_{1}$ is outputted.
 - $d\times d$ cheap operation, i.e. depth-wise convolution, is performed on $x_{1}$, and $x_{2}$ is outputted.
 - Then, $x_{1}$ and $x_{2}$ are concatenated.
 - $s$ is the ratio as above for generating $m=\frac{n}{s}$ intrinsic feature maps.

**Comment by NorbertZheng** [comment]
### Ghost Bottleneck
![image](https://user-images.githubusercontent.com/37235023/229414903-3909d917-9b04-4d55-931c-8137c89c2294.png)
Ghost bottleneck. Left: Ghost bottleneck with $stride=1$. Right: Ghost bottleneck with $stride=2$.
 - The Ghost bottleneck appears to be similar to the basic residual block in [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8?source=post_page---------------------------). The proposed ghost bottleneck mainly consists of two stacked Ghost modules.
 - The first Ghost module acts as an expansion layer increasing the number of channels. The ratio is the expansion ratio.
 - The second Ghost module reduces the number of channels to match the shortcut path.
 - When $stride=2$, the shortcut path is implemented by a downsampling layer and a depthwise convolution with $stride=2$ is inserted between the two Ghost modules.

**Comment by NorbertZheng** [comment]
### GhostNet
![image](https://user-images.githubusercontent.com/37235023/229415330-178c99e2-18bc-44b0-a25f-9ef3430de90f.png)
Overall architecture of GhostNet.

 - The model basically follows the architecture of [MobileNetV3](https://sh-tsang.medium.com/paper-mobilenetv3-searching-for-mobilenetv3-image-classification-5072d4d8703c), replacing the bottleneck block in [MobileNetV3](https://sh-tsang.medium.com/paper-mobilenetv3-searching-for-mobilenetv3-image-classification-5072d4d8703c) with our Ghost bottleneck.
 - All the Ghost bottlenecks are applied with $stride=1$ except that the last one in each stage is with $stride=2$.
 - The squeeze and excite (SE) module, as in [SENet](https://towardsdatascience.com/review-senet-squeeze-and-excitation-network-winner-of-ilsvrc-2017-image-classification-a887b98b2883?source=post_page---------------------------), is also applied to the residual layer in some ghost bottlenecks.
 - Yet, hard-[Swish](https://sh-tsang.medium.com/review-swish-searching-for-activation-functions-993a9ef2b4b9) nonlinearity function as in [MobileNetV3](https://sh-tsang.medium.com/paper-mobilenetv3-searching-for-mobilenetv3-image-classification-5072d4d8703c) is NOT used.
 - A width multiplier $\alpha$ is applied to scale the number of channels uniformly at each layer. GhostNet with width multiplier $\alpha$ as GhostNet- $\alpha\times$.

**Comment by NorbertZheng** [comment]
## Results
### Ablation Study
![image](https://user-images.githubusercontent.com/37235023/229416133-be5d2966-c9de-4a3b-a3b0-243019f55114.png)
The performance of the proposed Ghost module with different $d$ on CIFAR-10.
 - $s$ is used for generating $m=\frac{n}{s}$ intrinsic feature maps, and kernel size $d\times d$ of linear operations (i.e. the size of depthwise convolution filters) for calculating ghost feature maps.
 - $s=2$ and $d$ is tuned in $\{1,3,5,7\}$.
 - $d=3$ is the best.

![image](https://user-images.githubusercontent.com/37235023/229416567-0888b0a9-9184-44aa-aeda-c45e8afcec05.png)
The performance of the proposed Ghost module with different $s$ on CIFAR-10.
 - $d=3$ and $s$ is tuned in the range of $\{2,3,4,5\}$.
 - Larger $s$ leads to larger compression and speed-up ratio.
 - When $s=2$ which means compress [VGG](https://medium.com/coinmonks/paper-review-of-vggnet-1st-runner-up-of-ilsvlc-2014-image-classification-d02355543a11?source=post_page---------------------------)-16 by $2\times$, GhostNet performs even slightly better than the original model.

**Comment by NorbertZheng** [comment]
### CIFAR-10
![image](https://user-images.githubusercontent.com/37235023/229417182-7049654a-ae4b-41e1-9c3b-8b54db4bdf3a.png)
Comparison of state-of-the-art methods for compressing [VGG](https://medium.com/coinmonks/paper-review-of-vggnet-1st-runner-up-of-ilsvlc-2014-image-classification-d02355543a11?source=post_page---------------------------)-16 and [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8?source=post_page---------------------------)-56 on CIFAR-10.
 - Using ghost modules obtain small model size while keeping the accuracy.

**Comment by NorbertZheng** [comment]
### ImageNet
![image](https://user-images.githubusercontent.com/37235023/229417416-b9345420-ea0b-4b5e-9571-afead4768389.png)
Comparison of state-of-the-art methods for compressing [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8?source=post_page---------------------------)-50 on ImageNet dataset.
 - [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8?source=post_page---------------------------)-50 has about 25.6M parameters and 4.1B FLOPs with a top-5 error of 7.8%.
 - Ghost-[ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8?source=post_page---------------------------)-50 ($s=2$) obtains about $2\times$ acceleration and compression ratio, while maintaining the accuracy as that of the original [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8?source=post_page---------------------------)-50.

![image](https://user-images.githubusercontent.com/37235023/229417561-bba0c3b8-34f7-4080-b254-e9267a08bac8.png)
Comparison of state-of-the-art small networks over classification accuracy, the number of weights and FLOPs on ImageNet dataset.

![image](https://user-images.githubusercontent.com/37235023/229417696-25952fad-c274-4a49-8166-df5e555fe8c8.png)
Top-1 accuracy v.s. FLOPs/Latency on ImageNet dataset.

 - GhostNet obtain about 0.5% higher top-1 accuracy than [MobileNetV3](https://sh-tsang.medium.com/paper-mobilenetv3-searching-for-mobilenetv3-image-classification-5072d4d8703c) with the same latency, and GhostNet need less runtime to achieve similar performance.
 - For example, GhostNet with 75.0% accuracy only has 40 ms latency, while [MobileNetV3](https://sh-tsang.medium.com/paper-mobilenetv3-searching-for-mobilenetv3-image-classification-5072d4d8703c) with similar accuracy requires about 45 ms to process one image.
 - Overall, GhostNets generally outperform the famous state-of-art models, i.e. [MobileNetV2](https://towardsdatascience.com/review-mobilenetv2-light-weight-model-image-classification-8febb490e61c?source=post_page---------------------------), [MobileNetV3](https://sh-tsang.medium.com/paper-mobilenetv3-searching-for-mobilenetv3-image-classification-5072d4d8703c), [ProxylessNAS](https://sh-tsang.medium.com/paper-proxylessnas-direct-neural-architecture-search-on-target-task-image-classification-73c35ebd8aed), [FBNet](https://sh-tsang.medium.com/paper-fbnet-hardware-aware-efficient-convnet-design-via-differentiable-neural-architecture-f95a20c1cd4a), and [MnasNet](https://sh-tsang.medium.com/reading-mnasnet-platform-aware-neural-architecture-search-for-mobile-image-classification-b042aaef66f7).

**Comment by NorbertZheng** [comment]
### MS COCO
![image](https://user-images.githubusercontent.com/37235023/229417851-0957f5a8-308f-41dc-bc48-430b32ff3dd7.png)
Results on MS COCO dataset.
 - Both the two-stage [Faster R-CNN](https://towardsdatascience.com/review-faster-r-cnn-object-detection-f5685cb30202?source=post_page---------------------------) with Feature Pyramid Networks ([FPN](https://towardsdatascience.com/review-fpn-feature-pyramid-network-object-detection-262fc7482610?source=post_page---------------------------)) and the one-stage [RetinaNet](https://towardsdatascience.com/review-retinanet-focal-loss-object-detection-38fba6afabe4?source=post_page---------------------------) frameworks are used.
 - With significantly lower computational costs, GhostNet achieves similar mAP with [MobileNetV2](https://towardsdatascience.com/review-mobilenetv2-light-weight-model-image-classification-8febb490e61c?source=post_page---------------------------) and [MobileNetV3](https://sh-tsang.medium.com/paper-mobilenetv3-searching-for-mobilenetv3-image-classification-5072d4d8703c).

**Comment by NorbertZheng** [comment]
GhostNetV2 is then proposed in 2022 NeurIPS. Hope I can review it later.
